File: solutions/cpp/triangle/1/triangle.cpp
#include "triangle.h"

#include<stdexcept>

namespace triangle {

bool is_equilateral(double a, double b, double c)  {
    return (a == b) && (b == c);    
}

bool is_isosceles (double a, double b, double c) {
    return (a == b) || (a == c) || (b == c);    
}

bool is_scalene (double a, double b, double c) {
    return (a !=b) && (b!=c) && (c != a);    
}

bool satisfies_triangle_eq(double a, double b, double c){
   return (a+b >= c) && (b+c >= a) && (a+c >=b);
}

bool is_valid(double a, double b, double c){
    return (a>0) && (b>0) && (c>0) && satisfies_triangle_eq(a,b,c);
}

flavor kind(double a, double b, double c) {
    if (!is_valid(a,b,c)) 
        throw std::domain_error("Invalid triangle.");
    else {
        if (is_equilateral(a,b,c)) 
            return flavor::equilateral;
        else if (is_isosceles(a,b,c))
            return flavor::isosceles;
        else
            return flavor::scalene;
    }
}


}  // namespace triangle

File: solutions/free-pascal/triangle/1/Triangle.pas
unit Triangle;

{$mode ObjFPC}{$H+}

interface

function equilateral(const a, b, c : double): boolean;
function   isosceles(const a, b, c : double): boolean;
function     scalene(const a, b, c : double): boolean;

implementation

uses SysUtils;

function isvalid(const a,b,c : double) : boolean;
begin
  result := (a > 0) and (b > 0) and (c > 0);
  result := result and ((a+b>c) and (b+c>a) and (c+a>b));
end;

function equilateral(const a, b, c : double): boolean;
begin
  result:= isvalid(a,b,c) and ((a=b) and (b=c));
end;

function isosceles(const a, b, c : double): boolean;
begin
   result:= isvalid(a,b,c) and ((a=b) or (b=c) or (a=c));
end;

function scalene(const a, b, c : double): boolean;
begin
  result:= isvalid(a,b,c) and ((a<>b) and (b<>c) and (c<>a));
end;

end.
